src/systems/filecoin_vm/interpreter/vm_interpreter.go

```diff
@@ -124,51 +182,33 @@ func (vmi *VMInterpreter_I) ApplyMessage(inTree st.StateTree, message msg.Unsign
 		),
 	)
 
-	if !sendRet.ExitCode().AllowsStateUpdate() {
-		// error -- revert all state changes -- ie drop updates. burn used gas.
-		outTree = inTree
-		outTree = _withTransferFundsAssert(
-			outTree,
-			message.From(),
-			addr.BurntFundsActorAddr,
-			_gasToFIL(sendRet.GasUsed(), message.GasPrice()),
-		)
-	} else {
-		// success -- refund unused gas
-		outTree = sendRetStateTree
-		refundGas := message.GasLimit() - sendRet.GasUsed()
-		TODO() // TODO: assert refundGas is nonnegative
-		outTree = _withTransferFundsAssert(
-			outTree,
-			addr.BurntFundsActorAddr,
-			message.From(),
-			_gasToFIL(refundGas, message.GasPrice()),
-		)
+	ok = _vmiUseGas(sendRet.GasUsed())
+	if !ok {
+		panic("Interpreter error: runtime execution used more gas than provided")
```

Comment:
This is not a panic, this is just out of gas (exceeded the pre-set limit). Before execution, it's impossible to know how much will be actually consumed. An efficient implementation will of course halt execution before this point, but IMO the spec doesn't need to. In any case, the semantics are well defined: pay all the gas and revert state changes.

Reply:
It should be impossible for the runtime execution to use _more_ gas than provided -- as specced, it uses `SubtractWhileNonnegative` internally, and bails as soon as the gas usage would exceed the limit (without actually exceeding it).